src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -171,31 +171,22 @@ class ChannelData {
                                       grpc_connectivity_state* state,
                                       grpc_closure* on_complete,
                                       grpc_closure* watcher_timer_init) {
-    auto* watcher = new ExternalConnectivityWatcher(
+    auto watcher = new ExternalConnectivityWatcher(
         this, pollent, state, on_complete, watcher_timer_init);
     {
       MutexLock lock(&external_watchers_mu_);
-      // Will be deleted when the watch is complete.
-      GPR_ASSERT(external_watchers_[on_complete] == nullptr);
-      external_watchers_[on_complete] = watcher;
+      // Store a ref to the watcher in the external_watchers_ map.
+      watcher->AddWatcherToExternalWatchersMapLocked(&external_watchers_,
+                                                     on_complete);
     }
+    // Pass the ref from creating the object to Start().
     watcher->Start();
   }
 
   void RemoveExternalConnectivityWatcher(grpc_closure* on_complete,
                                          bool cancel) {
-    ExternalConnectivityWatcher* watcher = nullptr;
-    {
-      MutexLock lock(&external_watchers_mu_);
-      auto it = external_watchers_.find(on_complete);
-      if (it != external_watchers_.end()) {
-        watcher = it->second;
-        external_watchers_.erase(it);
-      }
-    }
-    // watcher->Cancel() will hop into the WorkSerializer, so we have to unlock
-    // the mutex before calling it.
-    if (watcher != nullptr && cancel) watcher->Cancel();
+    ExternalConnectivityWatcher::RemoveWatcherFromExternalWatchersMap(
+        &external_watchers_mu_, &external_watchers_, on_complete, cancel);
```

Comment:
Instead of passing in pointers to `external_watchers_mu_` and `external_watchers_`, just pass in `this`.  The `ExternalConnectivityWatcher` code can access the private members of `ChannelData`, because it's nested inside of it.

Reply:
Done!